internal/controller/pod_controller.go

```diff
@@ -81,51 +84,68 @@ func (a *PodAdmissionWebhook) Handle(ctx context.Context, req admission.Request)
+	wlConfigErr := a.updater.ConfigureWorkload(wl, proxies)
+
+	if wlConfigErr != nil {
+		l.Error(wlConfigErr, "Unable to reconcile workload result in webhook: "+wlConfigErr.Error(),
+			"kind", wl.Pod.Kind, "ns", wl.Pod.Namespace, "name", wl.Pod.Name)
+		return nil, fmt.Errorf("there is an AuthProxyWorkloadConfiguration error reconciling this workload %v", wlConfigErr)
+	}
+
+	return wl.Pod, nil // updated
+}
+
+// findMatchingProxies lists all AuthProxyWorkloads that are related to this pod
+// or its owners.
+func findMatchingProxies(ctx context.Context, c client.Client, u *workload.Updater, wl *workload.PodWorkload) ([]*cloudsqlapi.AuthProxyWorkload, error) {
 	var (
-		instList    = &cloudsqlapi.AuthProxyWorkloadList{}
-		proxies     []*cloudsqlapi.AuthProxyWorkload
-		wlConfigErr error
-		l           = logf.FromContext(ctx)
-		wl          = &workload.PodWorkload{Pod: &p}
+		instList = &cloudsqlapi.AuthProxyWorkloadList{}
+		proxies  []*cloudsqlapi.AuthProxyWorkload
+		l        = logf.FromContext(ctx)
 	)
 
 	// List all the AuthProxyWorkloads in the same namespace.
 	// To avoid privilege escalation, the operator requires that the AuthProxyWorkload
 	// may only affect pods in the same namespace.
-	err := a.Client.List(ctx, instList, client.InNamespace(wl.Object().GetNamespace()))
+	err := c.List(ctx, instList, client.InNamespace(wl.Object().GetNamespace()))
 	if err != nil {
 		l.Error(err, "Unable to list CloudSqlClient resources in webhook",
 			"kind", wl.Pod.Kind, "ns", wl.Pod.Namespace, "name", wl.Pod.Name)
 		return nil, fmt.Errorf("unable to list AuthProxyWorkloads, %v", err)
 	}
 
 	// List the owners of this pod.
-	owners, err := a.listOwners(ctx, wl.Object())
+	owners, err := listOwners(ctx, c, wl.Object())
 	if err != nil {
 		return nil, fmt.Errorf("there is an AuthProxyWorkloadConfiguration error reconciling this workload %v", err)
 	}
 
 	// Find matching AuthProxyWorkloads for this pod
-	proxies = a.updater.FindMatchingAuthProxyWorkloads(instList, wl, owners)
+	proxies = u.FindMatchingAuthProxyWorkloads(instList, wl, owners)
 	if len(proxies) == 0 {
 		return nil, nil // no change
 	}
 
-	// Configure the pod, adding containers for each of the proxies
-	wlConfigErr = a.updater.ConfigureWorkload(wl, proxies)
-
-	if wlConfigErr != nil {
-		l.Error(wlConfigErr, "Unable to reconcile workload result in webhook: "+wlConfigErr.Error(),
-			"kind", wl.Pod.Kind, "ns", wl.Pod.Namespace, "name", wl.Pod.Name)
-		return nil, fmt.Errorf("there is an AuthProxyWorkloadConfiguration error reconciling this workload %v", wlConfigErr)
-	}
+	return proxies, nil
 
-	return wl.Pod, nil // updated
 }
 
 // listOwners returns the list of this object's owners and its extended owners.
 // Warning: this is a recursive function
-func (a *PodAdmissionWebhook) listOwners(ctx context.Context, object client.Object) ([]workload.Workload, error) {
+func listOwners(ctx context.Context, c client.Client, object client.Object) ([]workload.Workload, error) {
```

Comment:
This refactors `listOwners()` from a member of `PodAdmissionWebhook` to a plain function so that it can be used both in the webhook and the pod event listener.